From 84260bf52fa9ed08ab1149b477f376d2519fe2f0 Mon Sep 17 00:00:00 2001
From: Imanol Fernandez <ifernandez@igalia.com>
Date: Fri, 28 Jan 2022 15:52:00 +0100
Subject: [PATCH 03/16] Disable ogg sandboxing

---
 toolkit/moz.configure | 1 -
 1 file changed, 1 deletion(-)

diff --git a/toolkit/moz.configure b/toolkit/moz.configure
index b76a1c9448ef..a8c0c4366801 100644
--- a/toolkit/moz.configure
+++ b/toolkit/moz.configure
@@ -2373,7 +2373,6 @@ set_define("MOZ_HAS_REMOTE", has_remote)
 def wasm_sandboxing_libraries():
     return (
         "graphite",
-        "ogg",
         "hunspell",
         "expat",
         "woff2",
-- 
2.35.1

