diff --git a/dom/vr/XRInputSource.cpp b/dom/vr/XRInputSource.cpp
index 412206cdf1ef..a7aca8f73cb8 100644
--- a/dom/vr/XRInputSource.cpp
+++ b/dom/vr/XRInputSource.cpp
@@ -138,6 +138,20 @@ nsTArray<nsString> GetInputSourceProfile(gfx::VRControllerType aType) {
       id.AssignLiteral("generic-trigger-squeeze-thumbstick");
       profile.AppendElement(id);
       break;
+    case gfx::VRControllerType::YvrTouch:
+      id.AssignLiteral("yvr-touch");
+      profile.AppendElement(id);
+      id.AssignLiteral("generic-trigger-squeeze-thumbstick");
+      profile.AppendElement(id);
+      break;
+    case gfx::VRControllerType::YvrTouch2:
+      id.AssignLiteral("yvr-touch-v2");
+      profile.AppendElement(id);
+      id.AssignLiteral("yvr-touch");
+      profile.AppendElement(id);
+      id.AssignLiteral("generic-trigger-squeeze-thumbstick");
+      profile.AppendElement(id);
+      break;
     default:
       NS_WARNING("Unsupported XR input source profile.\n");
       break;
diff --git a/gfx/vr/VRDisplayClient.cpp b/gfx/vr/VRDisplayClient.cpp
index 7978b4460505..e9adb17d36d9 100644
--- a/gfx/vr/VRDisplayClient.cpp
+++ b/gfx/vr/VRDisplayClient.cpp
@@ -440,6 +440,24 @@ void VRDisplayClient::GamepadMappingForWebVR(
       aControllerState.numButtons = 3;
       aControllerState.numAxes = 2;
       break;
+    case VRControllerType::YvrTouch:
+    case VRControllerType::YvrTouch2: {
+      aControllerState.buttonPressed =
+          ShiftButtonBitForNewSlot(0, 3) | ShiftButtonBitForNewSlot(1, 0) |
+          ShiftButtonBitForNewSlot(2, 1) | ShiftButtonBitForNewSlot(3, 4) |
+          ShiftButtonBitForNewSlot(4, 5) | ShiftButtonBitForNewSlot(5, 6);
+      aControllerState.buttonTouched = ShiftButtonBitForNewSlot(0, 3, true) |
+                                       ShiftButtonBitForNewSlot(1, 0, true) |
+                                       ShiftButtonBitForNewSlot(2, 1, true) |
+                                       ShiftButtonBitForNewSlot(3, 4, true) |
+                                       ShiftButtonBitForNewSlot(4, 5, true) |
+                                       ShiftButtonBitForNewSlot(5, 6, true);
+      aControllerState.axisValue[0] = aControllerState.axisValue[2];
+      aControllerState.axisValue[1] = aControllerState.axisValue[3];
+      aControllerState.numButtons = 6;
+      aControllerState.numAxes = 2;
+      break;
+    }
     default:
       // Undefined controller types, we will keep its the same order.
       break;
diff --git a/gfx/vr/external_api/moz_external_vr.h b/gfx/vr/external_api/moz_external_vr.h
index fad11a2a0307..ffc9a8812aed 100644
--- a/gfx/vr/external_api/moz_external_vr.h
+++ b/gfx/vr/external_api/moz_external_vr.h
@@ -188,6 +188,8 @@ enum class VRControllerType : uint8_t {
   Pico4,
   MetaQuest3,
   MagicLeap2,
+  YvrTouch,
+  YvrTouch2,
   _end
 };
 
@@ -218,6 +220,8 @@ inline constexpr bool IsEnumCase<gfx::VRControllerType>(
     case gfx::VRControllerType::MetaQuest3:
     case gfx::VRControllerType::MagicLeap2:
     case gfx::VRControllerType::PicoNeo3:
+    case gfx::VRControllerType::YvrTouch:
+    case gfx::VRControllerType::YvrTouch2:
     case gfx::VRControllerType::_end:
       return true;
   }
