--- a/netwerk/protocol/http/HttpBaseChannel.cpp
+++ b/netwerk/protocol/http/HttpBaseChannel.cpp
@@ -116,6 +116,8 @@
 namespace mozilla {
 namespace net {
 
+static constexpr auto kCloudflareChallengeHost = "challenges.cloudflare.com"_ns;
+
 static bool IsHeaderBlacklistedForRedirectCopy(nsHttpAtom const& aHeader) {
   // IMPORTANT: keep this list ASCII-code sorted
   static nsHttpAtomLiteral const* blackList[] = {
@@ -548,6 +550,15 @@
     return NS_OK;
   }
 
+  {
+    nsAutoCString hostFromURI;
+    nsresult rv = nsContentUtils::GetHostOrIPv6WithBrackets(mURI, hostFromURI);
+    if (NS_SUCCEEDED(rv)) {
+      if (StringEndsWith(hostFromURI, kCloudflareChallengeHost))
+        return NS_OK;
+    }
+  }
+
   NS_ConvertUTF16toUTF8 utf8CustomUserAgent(customUserAgent);
   nsresult rv = SetRequestHeader("User-Agent"_ns, utf8CustomUserAgent, false);
   if (NS_FAILED(rv)) return rv;
