diff --git a/dom/vr/XRInputSource.cpp b/dom/vr/XRInputSource.cpp
index f3beef26d9..412206cdf1 100644
--- a/dom/vr/XRInputSource.cpp
+++ b/dom/vr/XRInputSource.cpp
@@ -110,6 +110,12 @@ nsTArray<nsString> GetInputSourceProfile(gfx::VRControllerType aType) {
       id.AssignLiteral("generic-trigger-squeeze-thumbstick");
       profile.AppendElement(id);
       break;
+    case gfx::VRControllerType::PicoNeo3:
+      id.AssignLiteral("pico-neo3");
+      profile.AppendElement(id);
+      id.AssignLiteral("generic-trigger-squeeze-thumbstick");
+      profile.AppendElement(id);
+      break;
     case gfx::VRControllerType::Pico4:
       id.AssignLiteral("pico-4");
       profile.AppendElement(id);
diff --git a/gfx/vr/VRDisplayClient.cpp b/gfx/vr/VRDisplayClient.cpp
index 36a7b5bb71..2a5af2df3b 100644
--- a/gfx/vr/VRDisplayClient.cpp
+++ b/gfx/vr/VRDisplayClient.cpp
@@ -420,6 +420,7 @@ void VRDisplayClient::GamepadMappingForWebVR(
       aControllerState.numAxes = 2;
       break;
     case VRControllerType::PicoNeo2:
+    case VRControllerType::PicoNeo3:
     case VRControllerType::Pico4:
       aControllerState.buttonPressed =
           ShiftButtonBitForNewSlot(0, 3) | ShiftButtonBitForNewSlot(1, 0) |
diff --git a/gfx/vr/external_api/moz_external_vr.h b/gfx/vr/external_api/moz_external_vr.h
index 1bb2cdd9f8..4c7405e441 100644
--- a/gfx/vr/external_api/moz_external_vr.h
+++ b/gfx/vr/external_api/moz_external_vr.h
@@ -151,6 +151,7 @@ enum class VRControllerType : uint8_t {
   PicoGaze,
   PicoG2,
   PicoNeo2,
+  PicoNeo3,
   Pico4,
   MetaQuest3,
   MagicLeap2,
-- 
2.39.2

