--- a/gfx/gl/SharedSurfaceEGL.h
+++ b/gfx/gl/SharedSurfaceEGL.h
@@ -115,6 +115,8 @@ class SurfaceFactory_SurfaceTexture final : public SurfaceFactory {
  public:
   explicit SurfaceFactory_SurfaceTexture(GLContext&);
 
+  bool SupportsCspaces() const override { return true; }
+
   virtual UniquePtr<SharedSurface> CreateSharedImpl(
       const SharedSurfaceDesc& desc) override {
     return SharedSurface_SurfaceTexture::Create(desc);
