diff --git a/mobile/android/modules/geckoview/GeckoViewContent.sys.mjs b/mobile/android/modules/geckoview/GeckoViewContent.sys.mjs
index fd339e3d88..c71a5325f7 100644
--- a/mobile/android/modules/geckoview/GeckoViewContent.sys.mjs
+++ b/mobile/android/modules/geckoview/GeckoViewContent.sys.mjs
@@ -210,7 +210,8 @@ export class GeckoViewContent extends GeckoViewModule {
         this.sendToAllChildren(aEvent, aData);
         break;
       case "GeckoView:SetActive":
-        this.browser.docShellIsActive = !!aData.active;
+        if (this.browser.docShellIsActive != !!aData.active)
+          this.browser.docShellIsActive = !!aData.active;
         break;
       case "GeckoView:SetFocused":
         if (aData.focused) {
