diff --git a/dom/base/nsGlobalWindowOuter.cpp b/dom/base/nsGlobalWindowOuter.cpp
index 97d9671758..8fe8b656ca 100644
--- a/dom/base/nsGlobalWindowOuter.cpp
+++ b/dom/base/nsGlobalWindowOuter.cpp
@@ -6076,6 +6076,7 @@ void nsGlobalWindowOuter::FinalClose() {
   // complete _two_ round-trips to the event loop before the call to
   // ReallyCloseWindow. This allows setTimeout handlers that are set after
   // FinalClose() is called to run before the window is torn down.
+  /*
   nsCOMPtr<nsPIDOMWindowInner> entryWindow =
       do_QueryInterface(GetEntryGlobal());
   bool indirect = entryWindow && entryWindow->GetOuterWindow() == this;
@@ -6084,6 +6085,7 @@ void nsGlobalWindowOuter::FinalClose() {
   } else {
     mHavePendingClose = true;
   }
+  */
 }
 
 void nsGlobalWindowOuter::ReallyCloseWindow() {
