diff --git a/dom/canvas/ClientWebGLContext.cpp b/dom/canvas/ClientWebGLContext.cpp
index c06235685e..6f7cec7f83 100644
--- a/dom/canvas/ClientWebGLContext.cpp
+++ b/dom/canvas/ClientWebGLContext.cpp
@@ -4404,7 +4404,8 @@ void ClientWebGLContext::TexImage(uint8_t funcDims, GLenum imageTarget,
     const auto& contextInfo = mNotLost->info;
 
     const auto fallbackReason = [&]() -> Maybe<std::string> {
-      auto fallbackReason = BlitPreventReason(level, offset, pi, *desc);
+      auto fallbackReason =
+          BlitPreventReason(level, offset, pi, *desc, Limits());
       if (fallbackReason) return fallbackReason;
 
       const bool canUploadViaSd = contextInfo.uploadableSdTypes[sdType];
diff --git a/dom/canvas/TexUnpackBlob.cpp b/dom/canvas/TexUnpackBlob.cpp
index 73746cca5e..94315324b0 100644
--- a/dom/canvas/TexUnpackBlob.cpp
+++ b/dom/canvas/TexUnpackBlob.cpp
@@ -660,7 +660,8 @@ bool TexUnpackImage::Validate(const WebGLContext* const webgl,
 
 Maybe<std::string> BlitPreventReason(const int32_t level, const ivec3& offset,
                                      const webgl::PackingInfo& pi,
-                                     const TexUnpackBlobDesc& desc) {
+                                     const TexUnpackBlobDesc& desc,
+                                     const Limits& limits) {
   const auto& size = desc.size;
   const auto& unpacking = desc.unpacking;
 
@@ -691,8 +692,12 @@ Maybe<std::string> BlitPreventReason(const int32_t level, const ivec3& offset,
     }();
     if (premultReason) return premultReason;
 
-    if (pi.format != LOCAL_GL_RGBA) {
-      return "`format` is not RGBA";
+    if (pi.format != LOCAL_GL_RGBA && pi.format != LOCAL_GL_RGB) {
+      return "`format` is not RGBA or RGB";
+    }
+
+    if (pi.format == LOCAL_GL_RGB && !limits.rgbColorRenderable) {
+      return "`format` is RGB, which is not color-renderable";
     }
 
     if (pi.type != LOCAL_GL_UNSIGNED_BYTE) {
@@ -724,8 +729,8 @@ bool TexUnpackImage::TexOrSubImage(bool isSubImage, bool needsRespec,
 
   // -
 
-  const auto reason =
-      BlitPreventReason(level, {xOffset, yOffset, zOffset}, pi, mDesc);
+  const auto reason = BlitPreventReason(level, {xOffset, yOffset, zOffset}, pi,
+                                        mDesc, tex->mContext->Limits());
   if (reason) {
     webgl->GeneratePerfWarning(
         "Failed to hit GPU-copy fast-path."
diff --git a/dom/canvas/TexUnpackBlob.h b/dom/canvas/TexUnpackBlob.h
index 8d4c472b9a..5317af83eb 100644
--- a/dom/canvas/TexUnpackBlob.h
+++ b/dom/canvas/TexUnpackBlob.h
@@ -43,7 +43,8 @@ struct DriverUnpackInfo;
 
 Maybe<std::string> BlitPreventReason(int32_t level, const ivec3& offset,
                                      const webgl::PackingInfo&,
-                                     const TexUnpackBlobDesc&);
+                                     const TexUnpackBlobDesc&,
+                                     const Limits& limits);
 
 class TexUnpackBlob {
  public:
diff --git a/dom/canvas/WebGLContextValidate.cpp b/dom/canvas/WebGLContextValidate.cpp
index e5197db807..e1ecd7c914 100644
--- a/dom/canvas/WebGLContextValidate.cpp
+++ b/dom/canvas/WebGLContextValidate.cpp
@@ -237,6 +237,8 @@ static webgl::Limits MakeLimits(const WebGLContext& webgl) {
         gl::GLContext::KHR_texture_compression_astc_hdr);
   }
 
+  limits.rgbColorRenderable = webgl.gl->IsRGBColorRenderable();
+
   if (webgl.IsWebGL2() ||
       limits.supportedExtensions[WebGLExtensionID::WEBGL_draw_buffers]) {
     gl.GetUIntegerv(LOCAL_GL_MAX_DRAW_BUFFERS, &limits.maxColorDrawBuffers);
diff --git a/dom/canvas/WebGLTypes.h b/dom/canvas/WebGLTypes.h
index 0d4a20adfa..6ce293178d 100644
--- a/dom/canvas/WebGLTypes.h
+++ b/dom/canvas/WebGLTypes.h
@@ -651,6 +651,7 @@ struct Limits final {
 
   // Exts
   bool astcHdr = false;
+  bool rgbColorRenderable = false;
   uint32_t maxColorDrawBuffers = 1;
   uint64_t queryCounterBitsTimeElapsed = 0;
   uint64_t queryCounterBitsTimestamp = 0;
diff --git a/gfx/gl/GLContext.h b/gfx/gl/GLContext.h
index e22c732b60..8d3dbcd25d 100644
--- a/gfx/gl/GLContext.h
+++ b/gfx/gl/GLContext.h
@@ -283,6 +283,7 @@ class GLContext : public GenericAtomicRefCounted, public SupportsWeakPtr {
   }
 
   bool HasPBOState() const { return (!IsGLES() || Version() >= 300); }
+  bool IsRGBColorRenderable() { return !IsGLES() || Version() >= 300; }
 
   /**
    * If this context is double-buffered, returns TRUE.
